### PR TITLE
Migrate App lib to @nation3/utils

### PR DESCRIPTION
**Pull request by gregfromstl**
This is a **HIGH RISK** merge. It affects the entire app and has an incredibly wide scope. There will likely be bugs to sort out. Please test thoroughly.

Make sure you have an up-to-date `.env` file.

Test all existing functionality. Passport features need to be tested on rinkeby.

### CODE REVIEW

**Comment by aahna-ashina** [comment]
@gregfromstl I think a code conflict in `ui/lib/passport-nft.ts` needs to be resolved first?

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "app",
+  "lockfileVersion": 2,
+  "requires": true,
+  "packages": {}
+}
```
Comment (aahna-ashina) [inline package-lock.json:6]: Is this file necessary?

**Review by aahna-ashina — CHANGES_REQUESTED** [review]
There is a merge conflict.

**Comment by aahna-ashina** [comment]
@gregfromstl This is a lot of file changes in one pull request.  Do you think it would be possible to handle the migration more gradually, for example by doing only one page at the time or something like that?  Or is it only possible to do everything at once?

**Comment by johnmark13** [comment]
Or, is it possible to break this into two PRs. One PR for everything up to the last commit which looks like bug fixes and then another PR for the last commit that introduces the utils change?

I may have trivialised how interlocked the commits are but for example the change `if (!nationBalance && !veNationBalance) return` looks like something that could have been done/tested/released immediately as a bug fix that is actually not related to this migration?

**Comment by luisivan** [comment]
Few things:
- Seems like it's using external RPC providers instead of injected, in case there's one that's injected. Because of it, I'm getting CORS errors and nothing works
- Connecting and disconnecting an account doesn't work (I see a `TODO` message logged in the console while trying to disconnect)

**Comment by gregfromstl** [comment]
> @gregfromstl This is a lot of file changes in one pull request. Do you think it would be possible to handle the migration more gradually, for example by doing only one page at the time or something like that? Or is it only possible to do everything at once?

We likely have to do everything at once. This uses a WagmiProvider from the utils package rather than directly within the app. If we kept any of the wagmi imports directly in the app, it would require a different version (had to rollback for the utils package), causing all kinds of issues.

**Comment by gregfromstl** [comment]
> Or, is it possible to break this into two PRs. One PR for everything up to the last commit which looks like bug fixes and then another PR for the last commit that introduces the utils change?
> 
> I may have trivialised how interlocked the commits are but for example the change `if (!nationBalance && !veNationBalance) return` looks like something that could have been done/tested/released immediately as a bug fix that is actually not related to this migration?

The last PR was incorporating changes from a commit to main that came after everything else was done. I added two new hooks to the utils package and was just re-adding that functionality to the migrated version of the app.

**Comment by gregfromstl** [comment]
> There is a merge conflict.

Resolved

**Comment by gregfromstl** [comment]
@aahna-ashina do you know what's going on with the CI build

**Comment by gregfromstl** [comment]
> Few things:
> 
> * Seems like it's using external RPC providers instead of injected, in case there's one that's injected. Because of it, I'm getting CORS errors and nothing works
> * Connecting and disconnecting an account doesn't work (I see a `TODO` message logged in the console while trying to disconnect)
Screen Shot 2022-07-29 at 10 13 17 AM
Injected is being used if you check the `connectors` file in `shared/utils`. How do you know injected isn't being used? I'm not getting any CORS errors.

Good catch with disconnect, I meant to add that

**Comment by gregfromstl** [comment]
@luisivan disconnect is fixed

**Comment by johnmark13** [comment]
On the whole I think reviewing a PR like this should be about the application of coding standards and sanity checking nothing screaming "wrong" at the reviewer and on that basis I would approve the PR.

I'd have very moderate concerns that there are functional changes mixed in with the migration as that will make it harder to understand if the migration or the functional change (albeit they all look like fixes) changes behavior of the app. A trivial example of this is `downloadPassURI` where the code would have used to explode but now calls the endpoint with an undefined value, which could cause a different undefined behaviour. (That URI construction should move anyway to facilitate testing and multiple passport types, def a separate issue 😉 ). It's also best not to be too dogmatic!

As it looks like everyone is looking at the PR though and testing it for themselves, I think I am not needed, congratulations on what looks like an entertaining journey 💯

**Comment by aahna-ashina** [comment]
> > @gregfromstl This is a lot of file changes in one pull request. Do you think it would be possible to handle the migration more gradually, for example by doing only one page at the time or something like that? Or is it only possible to do everything at once?
> 
> We likely have to do everything at once. This uses a WagmiProvider from the utils package rather than directly within the app. If we kept any of the wagmi imports directly in the app, it would require a different version (had to rollback for the utils package), causing all kinds of issues.

@gregfromstl Got it.  What if we upgrade the wagmi version in the app, so that the wagmi version is the same in the app and in the utils library, _before_ doing the migration?  Would that make things less complicated?

**Comment by aahna-ashina** [comment]
> @aahna-ashina do you know what's going on with the CI build

@gregfromstl I compared the two branches, and I think the CI is failing because your branch contains a `package-lock.json` file.  The CI only expects the `yarn.lock` file to be present, which is the case on the main branch.

Maybe you can try using `yarn add <package>` instead of `npm install <package>`, and then delete the `package-lock.json` from your branch?

For context, see https://github.com/nation3/app/commit/eec0d97640e4db3c36054292c3670242243de95a and https://discord.com/channels/690584551239581708/963778898389237800/984017995913379871

**Review by aahna-ashina — CHANGES_REQUESTED** [review]
I suggest deleting `package-lock.json`, assuming we want to use Yarn instead of NPM.

**Comment by gregfromstl** [comment]
> > > @gregfromstl This is a lot of file changes in one pull request. Do you think it would be possible to handle the migration more gradually, for example by doing only one page at the time or something like that? Or is it only possible to do everything at once?
> > 
> > 
> > We likely have to do everything at once. This uses a WagmiProvider from the utils package rather than directly within the app. If we kept any of the wagmi imports directly in the app, it would require a different version (had to rollback for the utils package), causing all kinds of issues.
> 
> @gregfromstl Got it. What if we upgrade the wagmi version in the app so that the wagmi version is the same in the app and in the utils library, _before_ doing the migration? Would that make things less complicated?

The package wagmi version is a *rollback*. We can't use the latest version of wagmi since it breaks npm's dependency tree when used as a subdependency. Now, we could rollback the version of wagmi in the app, then replace the wagmi calls with nation3/utils ones piece by piece. But that would still require app-wide changes to update the wagmi version.

**Comment by luisivan** [comment]
@gregfromstl are you planning on re-opening this PR later?

**Pull request closed without merging**